### DISCUSSION TITLE
Splitting repos

### DISCUSSION THREAD

**Discussion by ericvergnaud** [WebAssembly/binaryen#6225] [General]
Following up on #6192 and #6224, that are both trying, in opposite ways, to address the issue of the TypeScript API not being in sync with the JavaScript API, itself not being in sync with the C++ API.

The opposite ways are:
 - move TypeScript wrapper from Binaryen.js to Binaryen and use it for generating the JavaScript wrapper, provide missing endpoints (#6192)
 - move the wrappers away completely from Binaryen, and host them in dedicated repos (such as Binaryen.js) (#6224)

What appears from various comments is 2 things:
 - _this_ repo's primary concern seems to be the C++ API (then why even generate WebAssembly?)
 - Binaryen.js lives in AssemblyScript org, but the AS team doesn't use it... 

It smells like a perfect setup for an unmaintained API... despite the talent and goodwill of maintainers

Do we need a different governance, where the repo for the JS wrapper of the WebAssembly of the C++ API is managed by its users rather than authors who have little or no interest in it ?

To clarify the goals of such a repo:
 - generate a lean WebAssembly that's in sync with the latest Binaryen C++ release (that can be used as is by wrapper authors)
 - provide a TypeScript API wrapping the above WebAssembly
 - provide or generate the corresponding documentation
 - package it and publish for use from node, deno and browsers

Ideally, that repo would live in the WebAssembly org, but I suspect it might take months to reach consensus on that... 
I'll happily create that repo (and transfer it later to WebAssembly org) if users are supportive...

**Comment by ericvergnaud** [comment]
An example of the TS/JS unsync is `const dataref: Type;` in index.d.ts, which does not exist in binaryen.js-post.js.
An example of the C++/JS unsync: TypeBuilder was released with v110 in Aug 22, but is not available with v116 of the JS API (proposed fixed by #6192 WIP).

**Comment by kripken** [comment]
> * this repo's primary concern seems to be the C++ API (then why even generate WebAssembly?)
> * Binaryen.js lives in AssemblyScript org, but the AS team doesn't use it...

Agreed that we are in an awkward position atm.

As background, we generate js/wasm here for the same reason e.g. the wabt repo does: it was easy to add, and in the same repo it's easy to keep in sync.

I would personally be ok either moving the JS from here to another repo, or moving the JS fully here. Both have advantages.

I do think [automation](https://github.com/WebAssembly/binaryen/pull/6192#issuecomment-1899053050) of the C and JS APIs might be a possible reason to do everything in this repo, but that would require some investigation probably.

**Reply by ericvergnaud** [reply]
Whilst I agree that partial automation is feasible, many critical endpoints require some coding for converting JS structures into something consumable by the WebAssembly: strings, arrays...

**Reply by ericvergnaud** [reply]
As I understand it, https://github.com/WebAssembly/binaryen/blob/main/src/wasm-delegations-fields.def currently implements 94 bindings. Binaryen.ts (from the PR) has 964 occurrences of "Binaryen" so probably 950+ bindings. That's quite a massive gap...

**Reply by tlively** [reply]
The wasm-delegations-fields.def file just gives the name and type of each field in each expression class. If we generated APIs from it, those would include getters and setters for each field as well as constructors for each expression class, so it would have much greater coverage than those numbers suggest. There would still be a gap for things like types, TypeBuilder, opcodes, and operations like running passes, parsing, and emitting modules, though.

**Comment by ericvergnaud** [comment]
Hi there. Any progress on the thinking ? I believe status quo is the worst option.

**Reply by kripken** [reply]
Personally I lean towards moving binaryen.js bindings to this repo because I think the automated bindings idea is promising, and would be best done in a single repo. (In theory we could wait on an experiment to validate that idea, if we are uncertain - but I am not worried myself.) But I don't know much about the buildbot side of things for binaryen.js. If it's hard to move that over, maybe it's not worth it.

And before any decision we need to make sure we have asked all the relevant people. In particular the people active on the binaryen.js stuff on the AssemblyScript repo. @dcodeIO , any concerns or thoughts? And is there anyone else we should ask?

**Reply by ericvergnaud** [reply]
Thanks for this feedback. Just summing up on the state of things, forgive my mistakes if any.

Re people active on the binaryen.js repo, iirc 25 were contacted (all contributors), 7 have responded. @ospencer is in favor of a split (PR #6224), whilst @csjh, @lbguilherme, @jayphelps, @dcodeIO plus myself seem to be in favor of a merge (PR #6192). @CountBleck is fine with both. Maybe there are other relevant people, but they haven't expressed any opinion that I'm aware of.

Re this repo, 3 people have expressed an opinion. @phated is in favor of a split. My understanding is that @kripken and @tlively lean towards a merge. Are there any people besides them whom should be involved ? 

As a reminder PR #6192 generates a lean wasm, which I believe satisfies the technical need of supporters of the split (but maybe not their governance concerns).
 
Re the binaryen.js buildbot, it comprises (@dcodeio let me know if I'm missing something):
 - a very thin wrapper ([3 lines of code](https://github.com/AssemblyScript/binaryen.js/blob/main/src/index.js)) that would disappear
 - the [automation](https://github.com/AssemblyScript/binaryen.js/blob/main/.github/workflows/build.yml) that compiles the wrapper and publishes the package to npm
 - the ts wrapper (already migrated to this repo if PR #6192 is accepted). It's not used by AssemblyScript.
 - the docs, that we would want to generate automatically (which is not the case today in either repo, so not sure how its technical feasibility affects a decision that should be rather imho driven by governance concerns)    

Worth noting that if we go for a merge, then the WebAssembly org would be the one publishing to NPM (I don't think it currently publishes anything there ?)

**Reply by kripken** [reply]
To make sure: I assume you mean "merge" as in "merge to this repo" while "split" means remove all JS from this repo and do it all somewhere else, of which "somewhere else" could be one of several options?

I lean towards a merge into this repo. But we'd need to figure out the automation and the publishing side, and I know nothing about how that works. (I assume somehow a github action has a key that it uses to upload to npm?)

cc @sbc100 regarding the last question (not trying to add work for you @sbc100, but just wanted to ping you since you may have done related things and have an opinion).

**Reply by sbc100** [reply]
I've not done stuff with npm publishing from github actions, but I assume its an extremely common thing to what to do and therefore fairly straight forward.

**Reply by ericvergnaud** [reply]
Yes I mean "merge" as in "merge to this repo". The automation is indeed an extremely common thing, but someone with admin rights will need to get involved in order to create the npm account, retrieve a secret key, and secretly store the secret key in GH...

**Reply by kripken** [reply]
Ok, it sounds like more people prefer a merge than a split, but before we move forward we should see how big of a concern a merge would be for those that prefer a split. Please comment here if a merge would be a significant burden for you in some way.

Other than that, we'd also need a volunteer to do the work to do the merge. (Everything but the secrets part, which we'll need to figure out.)

**Reply by ericvergnaud** [reply]
As a reminder, technically speaking, a merge will provide the same benefits as a split, since the generated js wrapping the wasm will be as lightweight as can be (all the heavyweight logic sits in a separately generated file that imports that wrapper).

I can volunteer to do the merge, if someone can provide support, notably for generating the documentation.

**Reply by phated** [reply]
In reading the various feedback, I don't understand why TypeScript bindings belong in this project, but OCaml, Haskell, Rust, etc bindings don't.

As I've mentioned in the other thread, I've tried to contribute to the bindings inside this repository and been consistently blocked.

If JS/TS experts want to make consistent improvements to bindings, they should do it in a separate repository where they are the primary stakeholders.

As for causing problems, I'm about 95% sure that @ericvergnaud's solution is going to create a very significant amount of work for Grain's js_of_ocaml bindings since he doesn't have the arcane knowledge in making it work. Additionally, we don't want to take on additional tooling, like the typescript compiler, to get our stuff to work.

Making something that doesn't churn is the best solution for me, @ospencer, and the rest of the Grain team as we don't have the bandwidth to keep up with said churn happening in this repository.

**Reply by kripken** [reply]
@phated 

> I've tried to contribute to the bindings inside this repository and been consistently blocked.

Can you please elaborate on that? My memory isn't good enough to remember what you are referring to, sorry, and it sounds like something we need to fix before considering changes here.

**Reply by phated** [reply]
@kripken here's an example that was opened 2 years ago and is still sitting: https://github.com/WebAssembly/binaryen/pull/4751

More than a couple times, you've stated something like "I don't have enough expertise around the JS bindings" for changes/suggestions/feedback I've provided.

I feel as though my removal of bindings solves this problem as it leaves the JS implementation to its specific stakeholders.

**Reply by CountBleck** [reply]
> But again, I do agree with the point that language-specific bindings generally make sense in separate repos where they can focus on just that, and they can be fully tested and so forth. Perhaps one way forward is to aim to autogenerate the C bindings in this repo, and leave other language bindings to other repos. However, I think those other languages could also benefit from that automation, which I'm not sure would be easy to do outside this repo, but I'm not sure.

(from https://github.com/WebAssembly/binaryen/pull/6192#issuecomment-1899053050)

@kripken I really like the idea of automating the bindings, and I want to direct more attention to it.

I'm sure it'd be very useful to have one or more JSON/YAML/TOML/etc. files in this repo that describe all the information that's needed to make well-typed wrappers around the C API in `binaryen-c.h`, which could also be generated by the aforementioned files.

This would let anyone make their own bindings, by piggybacking off of those description files to generate C API wrappers (without worrying about them silently breaking), and then handcrafting higher-level interfaces as the creator of those bindings desires. Hence, this kind of automation would benefit more than `binaryen.js` (which would then absorb the current `binaryen.js-post.js` file, as JS bindings would no longer need to be located here).

I think this would be much better than merging the JS/TS bindings completely into this repo, where it's less likely to change when necessary, and where it's less likely to appease every user that requires some variation of such bindings.

@phated @dcodeIO thoughts?

**Reply by sbc100** [reply]
Sounds a little like we talking about SWIG here (https://www.swig.org/)

**Reply by CountBleck** [reply]
@sbc100 Never heard of it. Is it any good, or is this whole concept inherently problematic?

**Reply by sbc100** [reply]
SWIG has been around basically forever (since that late nineties I think) but I've not used it in more than a decade.  

I think its hard to create "generic" binding yes, but that also sounds like what is being discussed here, so maybe we can avoid re-inventing too much or creating new formats?

**Reply by kripken** [reply]
@phated 

> here's an example that was opened 2 years ago and is still sitting: https://github.com/WebAssembly/binaryen/pull/4751

Thanks for the details. Ok, looks like the status of #4751 is that there are open review comments but you ran out of time to continue work on it, and asked if someone else could volunteer, and no one has so far?

I can't personally help more there (I don't have time to work on it myself, and can't tell anyone else what to work on), but I'd say it isn't blocked but rather waiting for someone with time and interest to do the work?

> More than a couple times, you've stated something like "I don't have enough expertise around the JS bindings" for changes/suggestions/feedback I've provided.

Yeah, my knowledge of the JS ecosystem is limited, I'm afraid. So I defer on JS matters to those that know better. Perhaps I wasn't clear enough about this - sorry if so - but my point is that I am not the best person to review such changes, but that just means we need to find someone that is. Among the people cc'd on this discussion are surely people with that knowledge, in particular.

Or, if we really can't find someone, we can still land things so long as we test them very carefully. I really want good PRs not to be blocked because of such reasons.

@sbc100 @CountBleck 

About bindings, I think there are two levels:

1. At a low level, this repo has `wasm-delegations-fields.h` which describes in a declarative manner all the instructions and fields we have. Using that it is simple to automate processes that need to do something for each field, like write a C API for getters and setters for each field (handling the type properly, etc.). That is the main thing I am suggesting we do, and it would be pretty lightweight (no new bindings library).
2. There is also the higher-level question of how to make bindings for many languages at once, which SWIG is one solution for. I don't know much about it myself. But we'd need e.g. JS bindings using the Emscripten JS glue, which I doubt SWIG supports?
  2.1. Instead of SWIG, we could do JS and other language bindings using `wasm-delegations-fields.h` as well. It would be more work per language, but the bindings could be more optimal and idiomatic I expect.

Concretely, I'd like to automate the C API first and see how that goes before considering anything else.

**Reply by CountBleck** [reply]
SWIG looks like "C++ -> bindings code in X language". That sounds like it'd break down for something like binaryen.js, where Binaryen is compiled with Emscripten (instead of using N-API).

Ideally the format should be something that's easy for Binaryen and other projects (of any language) to work with, without introducing weird dependencies. I'm definitely not suggesting the use of some arcane, proprietary format that has an overcomplicated type system (at that point, parsing clang AST might be better :P).

Even something like `["BinaryenCallIndirect", "Expression", ["Module", "string", "Expression", "ExpressionList", "Index", "Type", "Type"]]` would work, although there's definitely room to make it prettier (and parameter names would be nice). I'm thinking of something like OpenAPI, but much simpler.

**Reply by CountBleck** [reply]
@kripken

> There is also the higher-level question of how to make bindings for many languages at once

I think that should be delegated to binaryen.js and such (that is, bindings should be kept separate from this repo). This is why I think the format should be simple, so anyone can create low-level wrappers around the C API with a fifty line script, and then focus on making high-level idiomatic bindings.

> using the Emscripten JS glue, which I doubt SWIG supports?

Looks like you beat me to it.

**Reply by kripken** [reply]
I wrote a small proof of concept of automatic C API generation: https://github.com/WebAssembly/binaryen/pull/6291 Pretty simple to do, I think, and it could save a lot of manual effort.

**Comment by ericvergnaud** [comment]
> In reading the various feedback, I don't understand why TypeScript bindings belong in this project, but OCaml, Haskell, Rust, etc bindings don't.

I guess the reason for having TypeScript in this project is because it's a better way to produce JS bindings. 
Which then raises the question of why JS bindings belong to this repo ?

I personally think all language bindings should sit under the WebAssembly org, and preferably within the binaryen repo, simply because that's the best way to keep things in sync. The purpose of this discussion is to understand how it affects governance. Notably, it might require evolving from a centralized governance to a decentralized one (where experts are in charge of approvals in specific sub-areas). If that's not acceptable, then let's not do it. 

> As for causing problems, I'm about 95% sure that @ericvergnaud's solution is going to create a very significant amount of work for Grain's js_of_ocaml bindings since he doesn't have the arcane knowledge in making it work. Additionally, we don't want to take on additional tooling, like the typescript compiler, to get our stuff to work.
> 
> Making something that doesn't churn is the best solution for me, @ospencer, and the rest of the Grain team as we don't have the bandwidth to keep up with said churn happening in this repository.

@phated I think you might want to have a closer look ? Precisely, my proposal produces a very lean wasm wrapper, almost the same as the one you proposed (check [this](https://github.com/WebAssembly/binaryen/pull/6192/files#diff-deca46c00b1d791925adb5248307932987b43656f323ca60ff7a6e3cdb9c763a) vs [this](https://github.com/WebAssembly/binaryen/blob/main/src/js/binaryen.js-post.js)) and I believe it wouldn't require any additional work or additional tooling. The typescript bindings live alongside, not within, and they are in no way mandatory.

**Reply by phated** [reply]
@ericvergnaud I don't have the bandwidth to review, but as I mentioned, there is a bunch of arcane knowledge for js_of_ocaml. Do you have a non-wasm (wasm-to-js) build? Is that build es5 only?

**Reply by ericvergnaud** [reply]
My proposal doesn't touch the JS_OF_CAML build specifics (such as EXPORT_ES6=0). What it does is move almost all of what's currently included by --post-js to a separate file (it might even be possible to drop the --post-js completely via export flags, but I haven't tried that). So in principle, when JS_OF_CAML is on, the build should produce both wasm and non-wasm very lean es5 files.

**Reply by ericvergnaud** [reply]
@phated is your thumbs up an in-principle approval ?

**Reply by phated** [reply]
No, it means I read and understand your message. I continue to believe that drawing an arbitrary line at TypeScript bindings being included in this repository is entirely wrong, but it doesn't seem to matter to the primary stakeholders.

**Reply by kripken** [reply]
@phated I think we shouldn't move forward with any major change (and I see repo changes as major) that you and/or Grain are strongly opposed to.

**Reply by ericvergnaud** [reply]
Sorry but my understanding is that @phated is in favor of a major change (removing JavaScript out of this repo) whilst the other proposal is also a major change (move TypeScript into this repo).
@kripken are you suggesting that we don't change anything to a situation that doesn't satisfy anybody ?

**Reply by CountBleck** [reply]
@ericvergnaud On the contrary, @kripken (from my interpretation) desires a solution that satisfies just about everybody, and that's admittedly difficult.

**Reply by ericvergnaud** [reply]
@phated then may I suggest that you actually try out my proposed solution? I expect that you might like it because it satisfies your ask for a lean wasm. And if that's the case, then do you really care where JS/TS is maintained (my understanding is that you're not using it) ?

**Reply by kripken** [reply]
@ericvergnaud 

> are you suggesting that we don't change anything to a situation that doesn't satisfy anybody ?

I think any major change should ideally have something close to consensus among the relevant people. At least we shouldn't have major objections to it. We are not that large a group of people here, so if even one or two people strongly object that means we are making things worse for a significant proportion of us, which would be unfortunate. The current situation is also not ideal, I agree, and that motivates a change, but let's aim for one that makes sense to everyone.

Let's keep discussing here, but another option might be to schedule a video chat between interested people. That sometimes helps move discussions like this forward in my experience.

**Reply by samestep** [reply]
If a video chat occurs, may I join as well? I'm using Binaryen in TypeScript to develop program transformations and am interested in this conversation.

**Reply by kripken** [reply]
@samestep Sure, sounds good.

I suggest we give it another week or so for discussion here (as there is at least one question waiting for a response atm), and if we can't reach consensus by then, then let's look into scheduling a video chat.

Meanwhile maybe all the people interested in such a chat can respond with the :eyes: emoji on this comment of mine. Feel free to do so if you are a user of Binaryen's TypeScript or JavaScript APIs, or are otherwise interested in this.

**Reply by kripken** [reply]
I've sent an email to those interested in the meeting. Let me know if you are interested and somehow did not get the email (I found emails from git commit logs, so hopefully those are up to date).

**Reply by kripken** [reply]
We had a meeting about this now - thanks everyone that attended!

It was very productive I think, and we have a general plan for next steps here:

* Get automated (as much as possible) bindings working with embind, here in the Binaryen repo. We will use embind's TypeScript support to generate both JS and TS bindings. This is how we would do things today if we started from scratch, so it makes sense to try out, given the Emscripten bindings improvements in recent years.
  * I will look into the automation part (see #6291) and may ask for help from @brendandahl on the embind details.
  * We should try to make these bindings as minimal as possible.
  * We may consider two versions, for the wasm file embedded in JS and not (flipping `SINGLE_FILE` in emcc).
* Once that is done we can evaluate the API. It will not be identical to the current one, and likely lower-level. It may then make sense to layer a friendlier JS API on top, perhaps in a separate repo like other bindings have been doing.
  * A place for such separate repos might make sense in a dedicated binaryen org (which I reserved now: https://github.com/binaryen, just in case). Another option might be in the official WebAssembly org (`WebAssembly/binaryen.ts` perhaps).
  * The `src/wasm-delegations*.def` files have a high-level description of binaryen internals in a C header (that will be used in our autogeneration). A version of that as JSON or such might make it useful for other bindings generators. I can generate that and @phated may have ideas for the format.
* When the new embind-based build is looking good, we should look into setting up builds that are published on `binaryen.js` on npm, migrating the package to that basically. That will take some CI work.

**Comment by GulgDev** [comment]
With the official statement from the Binaryen.js repo owner about the bindings [not being actively maintained](https://github.com/WebAssembly/binaryen/issues/7526#issuecomment-2819012439) anymore, this problem is becoming more and more important. The gap between AssemblyScript's Binaryen TypeScript bindings and the actual API is very noticeable.

What is the current state of the situation? It seems like @kripken has a plan, but all the PRs that aim to resolve this issue are not moving forwards. Is the main issue that there's no one that has a good enough experience with `embind`? Maybe we can do something simpler for now, if the automatic binding generation approach is not working?

**Reply by kripken** [reply]
As my embind experiments (https://github.com/WebAssembly/binaryen/discussions/6225) did not work out, I'm open to other approaches.

In particular I'd be ok with the more incremental approaches in the existing PRs, if people are still interested?

**Reply by samestep** [reply]
Just mentioning since I was involved in this thread earlier: I've switched from Binaryen to `wasm-tools`, so I no longer have a stake in this issue.

**Reply by GulgDev** [reply]
Why not go with @ericvergnaud 's approach to compile TS module in the main Binaryen repo?

**Reply by GulgDev** [reply]
@kripken What if we use Embind on some simpler API, like the C wrapper and then when we get TS declarations for that wrapper, we could wrap it again so that it is compatible with the original JS wrapper? That would provide great type safety and also won't require many changes. I think it could be a solid start for migrating to using embind.

**Reply by kripken** [reply]
I am open to @ericvergnaud 's approach, yeah, as we agreed [when we met last year](https://github.com/WebAssembly/binaryen/discussions/6225#discussioncomment-8954759), it makes sense to move things to this repo, including TS generation (now that my main embind attempt is not moving forward). @ericvergnaud are you still interested in that?

Separately, the idea to use embind on the C API sounds worth exploring as well. (I don't know offhand how well embind deals with C as opposed to C++ types.)

**Reply by GulgDev** [reply]
I've done some experiments with TypeScript generation and Embind, but I still don't know what the result should look like. Is the compatibility with the old JS API necessary? If yes, should the bridge part be implemented on TS side or on C++ side? If no, what the end API should look like?

TypeScript generation for C works out of the box without Embind. So I think it is reasonable to first build the C API with TS bindings and then build the wrapper separately. However, I don't know if this solution is viable in the long term.

It seems like every approach had already been tried without success: kripken opened a PR to use `embind`, but got stuck; ericvergnaud's TypeScript PR got abandoned. What should be done to finally resolve this issue and start developing a single working solution?

**Reply by kripken** [reply]
It would be nice to keep the current API as much as possible, but if it is more natural or significantly simpler to change the API, I am open to that.

> TypeScript generation for C works out of the box without Embind.

How are you doing that? What does the TS look like?

**Reply by ericvergnaud** [reply]
@kripken Yes I'm still interested.

**Reply by GulgDev** [reply]
> How are you doing that? What does the TS look like?

Just adding the `--emit-tsd` option works for me. The emitted TS file contains types for every `Binaryen*` function, e.g.:
```
_BinaryenTypeNone(): number;
...
_BinaryenBrOnSetRef(_0: number, _1: number): void;
```

**Reply by GulgDev** [reply]
> It would be nice to keep the current API as much as possible, but if it is more natural or significantly simpler to change the API, I am open to that.

I think it is possible to recreate the current API using C++, but it seems odd to wrap a C wrapper of a C++ library. I don't know if it is any better than just adding a high-level TypeScript post-factum.

**Reply by ericvergnaud** [reply]
FWIW the PR is still live: https://github.com/WebAssembly/binaryen/pull/6192

**Reply by ericvergnaud** [reply]
> > How are you doing that? What does the TS look like?
> 
> Just adding the `--emit-tsd` option works for me. The emitted TS file contains types for every `Binaryen*` function, e.g.:
> 
> ```
> _BinaryenTypeNone(): number;
> ...
> _BinaryenBrOnSetRef(_0: number, _1: number): void;
> ```

How does that look like with more complex data structures ?

**Reply by GulgDev** [reply]
> > > How are you doing that? What does the TS look like?
> > 
> > 
> > Just adding the `--emit-tsd` option works for me. The emitted TS file contains types for every `Binaryen*` function, e.g.:
> > ```
> > _BinaryenTypeNone(): number;
> > ...
> > _BinaryenBrOnSetRef(_0: number, _1: number): void;
> > ```
> 
> How does that look like with more complex data structures ?

The resulting TS declaration file consists entirely of functions that take and return numbers, because that's just how the C wrapper works from the user perspective. I can post the file here, but I doubt that it would be useful.

**Reply by kripken** [reply]
@ericvergnaud I apologize for that PR being stalled for so long. But as other options are now ruled out, hopefully we can make a quick decision on the path forward here.

@GulgDev Looks like embind only manages to use the `number` type on those C APIs. In @ericvergnaud 's PR, there are user-defined types (like `export type ExpressionRef = number;`), which seem pretty useful for TS users?

Btw, separately, I do still hope we can automate more of these bindings. @GulgDev you already removed some duplication for the `getExpressionInfo` stuff, which is great, and regardless of the approach we go with here, I will eventually get to trying more stuff in https://github.com/WebAssembly/binaryen/pull/6291 and https://github.com/WebAssembly/binaryen/pull/6469. But none of that should block us here.

**Reply by GulgDev** [reply]
> In @ericvergnaud 's PR, there are user-defined types (like `export type ExpressionRef = number;`), which seem pretty useful for TS users?

In fact, they are not as useful as you think. TS's type system is structural rather than nominal. It means that any aliases are pretty much equivalent to number.

Although, it is possible to brand type using a hack (you can find more about branding types in TypeScript on the internet).

**Reply by GulgDev** [reply]
I've made a PoC: https://github.com/WebAssembly/binaryen/pull/7553

**Reply by ericvergnaud** [reply]
@GulgDev Thanks for the POC but tbh without the generated TypeScript it's hard to judge. 
From what I can gather, it could be a useful starting point to #6192, replacing manually written declarations such as `JSModule['_BinaryenTypeNone'](); ` with generated ones, making the whole thing safer and easier to maintain ?

**Reply by GulgDev** [reply]
If you look at the source code, you can see that I used delegations to automate expression bindings. I am still working on it, and it's already far simpler and more optimal than the current JS wrapper. It also doesn't require any additional compile steps, making it faster to compile than TS wrapper. The resulting type defenitions are pretty accurate, as Embind tries to recreate the C++ types exactly.

**Reply by ericvergnaud** [reply]
Some binaryen consumers won't work with TS, so we still need to generate JS. 
The additional compile step is only required to generate JS from TS. 
That is true for both your proposal and #6192.

**Reply by GulgDev** [reply]
I don't get your point. It's pointless to generate TS without JS (or vise-versa). Embind generates both TS and JS at the same time.

**Reply by ericvergnaud** [reply]
Ah! That wasn't clear... Good to hear.
Anyways I'm not sure if compilation time is a concern.
Rather I'm looking for an expressive high-level API, that can be tested automatically each time there is a change in binaryen core. I suspect most developers also want that, rather than just minimal bindings to flat C functions.
There was one originally maintained as part of AssemblyScript, but that became gradually disconnected, which is what triggered this initiative.

**Reply by CountBleck** [reply]
(Note that the *typings* and the buildbot were maintained by AS, not the JS binding code.)

I agree that most users of binaryen.js (outside AssemblyScript) generally want a high-level API, so something pretty close to the C++ API would be nice.
